lib/Dialect/FIRRTL/Transforms/ExtractClasses.cpp

```diff
@@ -106,18 +219,30 @@ void ExtractClassesPass::extractClass(FModuleOp moduleOp) {
         cast<FIRRTLPropertyValue>(moduleOp.getArgument(outputProperty.index));
     Value originalValue = getDriverFromConnect(outputValue);
 
-    // If the Value is defined by an Operation that hasn't been copied yet, copy
-    // that into the body, and map from the old Value to the new Value. This may
-    // need to walk property ops in order to copy them into the ClassOp, but for
-    // now only constant ops exist. Mark the property op to be erased.
+    // If the Value is defined by an Operation that has already been processed,
+    // there is nothing to do.
     if (!mapping.contains(originalValue)) {
       if (auto *op = originalValue.getDefiningOp()) {
-        builder.clone(*op, mapping);
-
-        // InstanceOps are handled specially for now, but any other property
-        // defining ops should be erased after being copied over.
-        if (!isa<InstanceOp>(op))
+        // InstanceOps are handled specially, by creating ObjectOps and
+        // extracting an ObjectFieldOp. This will take care of re-using
+        // ObjectOps and ObjectFieldOps, updating the mapping, and keeping
+        // track of related ops that can be erased. The original value is
+        // mapped to the ObjectFieldOp.
+        if (auto instance = dyn_cast<InstanceOp>(op)) {
+          Value fieldValue =
+              getOrCreateObjectValue(cast<OpResult>(originalValue), instance,
```

Comment:
It looks like objects are only created from instances if their output properties drive the containing module's output properties. So instances with only input properties or unconnected output properties won't be represented as objects. Is that intentional?

Reply:
Hmm that's a good point. This pass is coming from the angle of if something isn't ultimately assigned to the containing module's output properties, it isn't "public" or visible. Let me add a test case where there are internal instances (for example an instance, whose output is an input to another instance, whose output is assigned to a module output). That kind of thing should work.

Reply:
Good call, I was missing support for this case... basically, getOrCreateObject needs to have the same special case handling of what to do when an instance is encountered as extractClass. I've pulled that code out into a common helper, and this case now works. clang-tidy is unhappy I've implemented this with mutually recursive calls, but I think this is reasonable for now.